### DISCUSSION TITLE
Cloud infra info and performance

### DISCUSSION THREAD

**Discussion by Buzut** [appwrite/appwrite#8676] [General]
First of all, thank you so much for building AppWrite and making it open source! I love the UI and the features.

Now I can't find any information on the current cloud infrastructure apart that it's hosted in Germany. Is this AWS/GCP or the likes under the hood? Redundancy?

Although the Cloud is still in Beta phase, before committing with a provider, I think it's important to know what we're going with.

So far, each time I give the Cloud a try, it's very slow and so I'm discouraged of using it for any serious project at this stage. 

This very simple db query (server side) is more than half a sec although there are less than 5 entries in the tables:

`https://cloud.appwrite.io/v1/databases/66a66c150013ad553b55/collections/66e92743002161025f62/documents`
```json
{
	"total": 2,
	"documents": [
		{
			"slug": "sr-lb",
			"parent_id": "",
			"noindex": false,
			"$id": "66e928f40038a1367242",
			"$createdAt": "2024-09-17T07:00:05.196+00:00",
			"$updatedAt": "2024-09-17T08:01:42.925+00:00",
			"$permissions": [],
			"tags": [
				{
					"tag": "power",
					"$id": "66e92c0b0038b4882bec",
					"$createdAt": "2024-09-17T07:13:16.034+00:00",
					"$updatedAt": "2024-09-17T07:13:16.034+00:00",
					"$permissions": [],
					"$databaseId": "66a66c150013ad553b55",
					"$collectionId": "66e92be20002cb4e8762"
				}
			],
			"$databaseId": "66a66c150013ad553b55",
			"$collectionId": "66e92743002161025f62"
		},
		{
			"slug": "sr-ub",
			"parent_id": "",
			"noindex": false,
			"$id": "66e92bbe0035d862f17d",
			"$createdAt": "2024-09-17T07:11:58.961+00:00",
			"$updatedAt": "2024-09-17T07:11:58.961+00:00",
			"$permissions": [],
			"tags": [],
			"$databaseId": "66a66c150013ad553b55",
			"$collectionId": "66e92743002161025f62"
		}
	]
}
```

Also just ran a very simple user/email session creation, slightly more than 500ms… And I'm in France, so I wouldn't say Germany is that far. In the meantime, I get sub 50ms for similar queries with other services.

![Screenshot 2024-09-18 at 06 19 18](https://github.com/user-attachments/assets/ec3c1ea0-769f-486f-bbf7-a705942b679f)

**Comment by DH-555** [comment]
Hello! I can't give detailed information on that, but Appwrite cloud has redundant servers.

More than 5 seconds seems to be unusual and slow for that action. What queries are you running for it?
Also, sorry for the ultra-slow response. I'm not very active here.

**Reply by Buzut** [reply]
It's not 5 seconds, it's half a sec, but 500ms is huge, doesn't pass web vitals at all. 
I explained all the queries in the post, first one is a simple GET on the DB through the server API, the second is a User/Email account creation. I tried through REST an JS SDK, same results, at several weeks of interval and from several locations in France.

No worries for the slow reply, we all have a lot on our tables ;)

**Reply by DH-555** [reply]
@Buzut You can't expect a 50ms response like in a standard database since the requests being made to the database pass through several things such as an API and Authentication checks, but 500ms is not expected or the usual latency.

Do you can join the Appwrite Discord server and post this in the Support channel?
Some member from the Appwrite core team will investigate with you those performance issues you're facing and try to fix it.
https://appwrite.io/discord

**Reply by Buzut** [reply]
Yeah I'll ping them on Discord. Apples to Apples, I just made the very same test with Supabase and the response time is order of magnitude lower. Now I don't like Supabase for different reasons so I'd rather use AppWrite ;)

**Comment by minecrawler** [comment]
just wanted to chip in, I had a DB query in my Cloud test project with less than 10 rows which took over 10 SECONDS to finish. I just wanted to get all fields of a row with a specific ID. The time fluctuates a lot, though, so it's hard to give detailed steps to reproduce it.

**Reply by DH-555** [reply]
That's weird. Do you can share if possible the queries you're running and the database structure?
Sharing the network logs will be helpful too. Remember to censor sensitive content.
Are you using relationships?

**Reply by minecrawler** [reply]
It's a very simple query for an index. No relationships. Just one table with a few attributes. Very straight forward. I'll share the logs when I see the problem next time. Unfortunately, it's fluctuating for me

**Reply by minecrawler** [reply]
@DH-555 here's a slow one ("only" 2.5s, though). It reads the first 25 entries from a database table and returns them ordered by creation date. There are currently 2 or 3 entries. As you can see it's not the only slow one. They all need about the same time, and each of them is doing a similar thing. Reading posts, dashboard widgets, and additional user data. Nothing fancy and on a WIP project.

There are also faster ones, so it's not just a slow connection on my end.

![image](https://github.com/user-attachments/assets/58c69405-2baf-40cc-a093-0f4308e2ce9b)

**Reply by DH-555** [reply]
@minecrawler From what I'm seeing, you're getting the documents through a function. Do you can give me the following information?

- What's the runtime that you're using?
- Are the times improving after the first function run?
- What requests and operations is the function doing? (for example: getting a document and making an operation)
- What are the queries being performed? (for example: [Query.equals("Person", "D5")])

**Comment by eldadfux** [comment]
Hey @Buzut, thank you so much for your kind words. I'll do my best the shed some clarity and help with your concerns.

To start, due to data privacy and security restrictions we can not disclose specific information about Cloud's infrastructure design, machines types, or vendor names, but we can share a good amount of details and you can also assume that the basic design is based in majority on the design of the open source edition with important optimization for Cloud (hyper) scale (Appwrite Cloud is by far the biggest Appwrite setup in the wild to the best of our knowledge).

Appwrite Cloud uses a multi-cloud architecture, meaning we utilize multiple cloud vendors simultaneously, including major cloud vendors. For each vendor, we strictly use only standard compute, DNS, storage, CDN services, and additional security layers like DDoS protection where relevant. We do not use any managed database services, all our databases are self-managed using high-availability setups.

This design avoids any form of vendor lock-in, both in terms of business agreements and software. It allows us to host Appwrite on any provider, which is especially crucial for our self-hosted edition and our enterprise offerings that includes “bring your own cloud” and must be able to work on any provider or locally on your laptop.

On top of that, all of Appwrite API services and workers are designed with highly available, distributed, and redundant setups. Scaling is easy and we take leverage of sharding and other techniques to easily scale horizontally or vertically. 

We’re continuously working to improve the performance of Appwrite, which has been a major focus for us for a long time. Version 1.6 introduced significant enhancements, and we’ve recently implemented more cloud-specific improvements that have shown over 70%-100% reduction in latency.

We’re actively working on additional ways to boost Appwrite’s performance as Cloud beta continues.  We expect Appwrite 1.7 to be our fastest version yet, with massive rewrites of our internal API engines showing over 100% improvements in lab environment benchmarks. We’re very excited about this and other planned improvements that are being released regularly.

If anyone encounters performance issues, our support team is more than happy to assist, and our engineers are also often available for advice in our Discord community.

I will also make sure to have this kind of information available on our official docs, I have added this to our backlog.

**Reply by Buzut** [reply]
Late to the party, but thank you for the detailed explanation Eldad!

**Comment by deepshekhardas** [comment]
Adding a self-hosted perspective that might help others comparing Cloud vs self-hosted performance:

We migrated a production workload off Appwrite Cloud to self-hosted after hitting similar latency issues (~400-600ms for simple document reads). Post-migration results on the same query:

**Cloud:** ~500ms avg (single-region, Germany)
**Self-hosted (Hetzner CX31, same Frankfurt region):** ~18-35ms avg

The delta was almost entirely **network + cold-start overhead** on the Cloud side, not the Appwrite engine itself.

**What helped on self-hosted:**

1. **MariaDB tuning** — increasing `innodb_buffer_pool_size` to 70% of available RAM made the biggest single improvement
2. **Connection pooling** — Appwrite uses its own internal pooling but if you're running high concurrency, make sure your `_APP_DB_POOL_SIZE` env var is set appropriately (default is quite low)
3. **Redis** — ensure Redis is on the same host or very low-latency network. Cache misses hit the DB hard on Appwrite
4. **Index your collections** — Appwrite's collection indexes map directly to DB indexes. Any query without an index = full scan, which explains 500ms+ on even small collections

```bash
# Check if your collection has indexes:
# Appwrite Console > Database > Collection > Indexes
# Always add indexes for fields used in queries/filters
```

5. For the Cloud specifically — if you're on the free tier, resources are shared. Pro tier showed noticeably better p99 latency in my testing.

If you're hitting this at enterprise scale and need the self-hosted setup properly optimized, I've handled similar migrations. I work via async sprints — feel free to connect on [LinkedIn](https://www.linkedin.com/in/deepshekhardas/) if you need this sorted fast.
